deliver/lib/deliver/upload_metadata.rb

```diff
@@ -226,30 +226,47 @@ def verify_available_languages!(options)
       # Collect all languages we need
       # We only care about languages from user provided values
       # as the other languages are on iTC already anyway
-      v = options[:app].edit_version
-      UI.user_error!("Could not find a version to edit for app '#{options[:app].name}', the app metadata is read-only currently") unless v
+      v = options[:app].details
+      UI.user_error!("Could not find details for app '#{options[:app].name}'") unless v
+
+      # find languages currently enabled on ITC (enabled_languages)
+      enabled_languages = {}  # map of locale code to language data hash
+      primary_data = nil      # the primary language data, e.g. english name/subtitle
+      v.languages.each do |language_data|
+        lang = language_data["localeCode"]
+        enabled_languages[lang] = language_data
+        primary_data = language_data unless language_data["canDeleteLocale"]
+      end
 
-      enabled_languages = options[:languages] || []
+      UI.user_error!("could not find primary language data for app '#{options[:app].name}'") unless primary_data
+
+      # find list of languages we have data for (local_languages)
+      local_languages = options[:languages] || []
       LOCALISED_VERSION_VALUES.each do |key|
         current = options[key]
         next unless current && current.kind_of?(Hash)
         current.each do |language, value|
-          language = language.to_s
-          enabled_languages << language unless enabled_languages.include?(language)
+          lang = language.to_s
+          next if lang == "default"
+          local_languages << lang unless local_languages.include?(lang)
         end
       end
 
-      # Reject "default" language from getting enabled
-      # because "default" is not an iTC language
-      enabled_languages = enabled_languages.reject do |lang|
-        lang == "default"
-      end.uniq
+      UI.verbose("ITC languages: #{enabled_languages.keys}")
+      UI.verbose("local languages: #{local_languages}")
+      # determine languages to add
+      add_languages = local_languages - enabled_languages.keys
+
+      if add_languages.count > 0
+        # we need to add languages to ITC
+        # update v.languages
+        add_languages.each do |locale|
+          v.languages << primary_data.merge("localeCode" => locale, "canDeleteLocale" => true)
```

Comment:
huh what does this do exactly? `primary_data` is unexpected here.

Reply:
`primary_data` is a hash with settings e.g. for english. this code just copies the primary data to the new language, and changes locale code to the target locale/language. the data is later overwritten by the normal fastlane upload process

Reply:
ah, so you use `primary_data` as a kind of "template" where you add/change two values for the language to be added, correct?

Reply:
yes that's right